homeassistant/components/auth/__init__.py

```diff
@@ -222,45 +224,45 @@ def __init__(self, retrieve_credentials):
         credentials = self._retrieve_credentials(
             data['client_id'], data['code'])
 
-        if credentials is None:
+        if credentials is None or not isinstance(credentials, Credentials):
             return self.json_message('Invalid code', status_code=400)
 
         await hass.auth.async_link_user(user, credentials)
         return self.json_message('User linked')
 
 
 @callback
-def _create_cred_store():
-    """Create a credential store."""
-    temp_credentials = {}
+def _create_auth_code_store():
+    """Create an in memory store."""
+    temp_results = {}
 
     @callback
-    def store_credentials(client_id, credentials):
-        """Store credentials and return a code to retrieve it."""
+    def store_result(client_id, result):
+        """Store flow result and return a code to retrieve it."""
         code = uuid.uuid4().hex
-        temp_credentials[(client_id, code)] = (dt_util.utcnow(), credentials)
+        temp_results[(client_id, code)] = (dt_util.utcnow(), result)
         return code
 
     @callback
-    def retrieve_credentials(client_id, code):
-        """Retrieve credentials."""
+    def retrieve_result(client_id, code):
```

Comment:
Let's make this `(client_id, type, code)`

Reply:
It is little hard for `save_result` to add type, since the caller `LoginFlowResourceView.post` don't have know what flow result type is expecting. I have to inspect its datatype to generate the `type`

Reply:
Ah you're right, we don't even have the context or flow 😢